4.x: antlr-runtime dependency convergence

## Environment Details
* Helidon Version: 4.0.1

----------

## Problem Description

```
[ERROR] Rule 0: org.apache.maven.enforcer.rules.dependency.DependencyConvergence failed with message:
[ERROR] Failed while enforcing releasability.
[ERROR] 
[ERROR] Dependency convergence error for org.antlr:antlr4-runtime:jar:4.9.3 paths to dependency are:
[ERROR] +-io.helidon:helidon-all:pom:4.0.0-SNAPSHOT
[ERROR]   +-io.helidon.graphql:helidon-graphql-server:jar:4.0.0-SNAPSHOT:compile
[ERROR]     +-com.graphql-java:graphql-java:jar:18.6:compile
[ERROR]       +-org.antlr:antlr4-runtime:jar:4.9.3:runtime
[ERROR] and
[ERROR] +-io.helidon:helidon-all:pom:4.0.0-SNAPSHOT
[ERROR]   +-io.helidon.integrations.cdi:helidon-integrations-cdi-hibernate:jar:4.0.0-SNAPSHOT:compile
[ERROR]     +-org.hibernate.orm:hibernate-core:jar:6.3.1.Final:compile
[ERROR]       +-org.antlr:antlr4-runtime:jar:4.10.1:runtime
```

## Steps to reproduce

Comment-out workaround in all/pom.xml (dependencyManagement at the bottom).

```bash
mvn validate -f all/pom.xml
```

---

The fix to this issue should remove the work-arounds in the following files (see #8171):
- `all/pom.xml`
- `dependencies/pom.xml`

I have tried fixing this by upgrading graphql-java to `22.1`, but then we have a problem with failing tests:
```
[ERROR] Failures:
[ERROR]   DateTimeIT.testDateAndTime:152
Expected: is "02/17/1968"
     but: was <1968-02-17>
[ERROR]   DateTimeIT.testDateInputsAsPojo:283->validateResult:325
Expected: is "02/17/1968"
     but: was <1968-02-17>
[ERROR]   MapIT.testMap:55->AbstractGraphQLTest.getAndAssertResult:218 Errors: Server Error

[ERROR]   NumberFormatIT.testNumberFormats:72
Expected: is "1 id"
     but: was "1"
[ERROR] Errors:
[ERROR]   DateTimeIT.testDatesAndMutations:243 ClassCast class java.time.LocalDate cannot be cast to class java.lang.String (java.time.LocalDate and java.lang.String are in module java.base of loader 'bootstrap')
[ERROR]   DateTimeIT.testDifferentSetterGetter:72 ClassCast class java.time.LocalDate cannot be cast to class java.lang.String (java.time.LocalDate and java.lang.String are in module java.base of loader 'bootstrap')
[ERROR]   DateTimeScalarIT.testDateAndTime:66 ClassCast class java.time.LocalDate cannot be cast to class java.lang.String (java.time.LocalDate and java.lang.String are in module java.base of loader 'bootstrap')
```

--- a/all/pom.xml
+++ b/all/pom.xml
@@ -1130,16 +1130,4 @@
         </dependency>
     </dependencies>
 
-    <dependencyManagement>
-        <dependencies>
-            <dependency>
-                <!--
-                    Work-around for https://github.com/helidon-io/helidon/issues/8172
-                 -->
-                <groupId>org.antlr</groupId>
-                <artifactId>antlr4-runtime</artifactId>
-                <version>4.10.1</version>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
 </project>

--- a/config/metadata/docs/pom.xml
+++ b/config/metadata/docs/pom.xml
@@ -103,21 +103,6 @@
         </dependency>
     </dependencies>
 
-    <dependencyManagement>
-        <dependencies>
-            <dependency>
-                <!--
-                    Work-around for https://github.com/helidon-io/helidon/issues/8172
-                    dependency convergence of antlr, where graphql uses an older version
-                    and fails when we upgrade antlr
-                 -->
-                <groupId>org.antlr</groupId>
-                <artifactId>antlr4-runtime</artifactId>
-                <version>4.10.1</version>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
-
     <build>
         <plugins>
             <plugin>

--- a/dependencies/pom.xml
+++ b/dependencies/pom.xml
@@ -55,8 +55,8 @@
         <version.lib.google-error-prone>2.3.3</version.lib.google-error-prone>
         <version.lib.google-protobuf>3.21.7</version.lib.google-protobuf>
         <version.lib.graalvm>23.1.0</version.lib.graalvm>
-        <version.lib.graphql-java>18.6</version.lib.graphql-java>
-        <version.lib.graphql-java.extended.scalars>18.3</version.lib.graphql-java.extended.scalars>
+        <version.lib.graphql-java>22.1</version.lib.graphql-java>
+        <version.lib.graphql-java.extended.scalars>22.0</version.lib.graphql-java.extended.scalars>
         <version.lib.gson>2.9.0</version.lib.gson>
         <version.lib.grpc>1.60.0</version.lib.grpc>
         <version.lib.guava>32.0.1-jre</version.lib.guava>

--- a/docs/pom.xml
+++ b/docs/pom.xml
@@ -88,19 +88,6 @@
         </dependency>
     </dependencies>
 
-    <dependencyManagement>
-        <dependencies>
-            <dependency>
-                <!--
-                    Work-around for https://github.com/helidon-io/helidon/issues/8172
-                 -->
-                <groupId>org.antlr</groupId>
-                <artifactId>antlr4-runtime</artifactId>
-                <version>4.10.1</version>
-            </dependency>
-        </dependencies>
-    </dependencyManagement>
-
     <build>
         <finalName>${project.artifactId}</finalName>
         <plugins>

--- a/graphql/server/src/main/java/io/helidon/graphql/server/InvocationHandlerImpl.java
+++ b/graphql/server/src/main/java/io/helidon/graphql/server/InvocationHandlerImpl.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2024 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -26,6 +26,8 @@ import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
 import graphql.ExceptionWhileDataFetching;
 import graphql.ExecutionInput;
@@ -49,6 +51,11 @@ import static io.helidon.graphql.server.GraphQlConstants.PATH;
 
 class InvocationHandlerImpl implements InvocationHandler {
     private static final System.Logger LOGGER = System.getLogger(InvocationHandlerImpl.class.getName());
+    private static final Pattern VALIDATION_ERROR_PATTERN = Pattern.compile("Validation error \\((\\w+)@\\[(.*)]\\) : (.*$)");
+    private static final Pattern ERROR_ENUM_REPLACE_PATTERN =
+            Pattern.compile("Literal value not in allowable values for enum '.*?' - ('.*\\{.*}')");
+    private static final Pattern INVALID_TYPE_REPLACE_PATTERN =
+            Pattern.compile("- Expected an AST type of ('.*?') but it was a ('.*?') (@.*?)$");
 
     private final String defaultErrorMessage;
     private final Set<String> exceptionDenySet = new HashSet<>();
@@ -201,7 +208,71 @@ class InvocationHandlerImpl implements InvocationHandler {
             path = listPath.get(0).toString();
         }
 
-        addErrorPayload(resultMap, error.getMessage(), path, line, column, error.getExtensions());
+        addErrorPayload(resultMap, fixMessage(error.getMessage()), path, line, column, error.getExtensions());
+    }
+
+    // this works around https://github.com/eclipse/microprofile-graphql/issues/520
+    // once the TCK is fixed, we can remove this work-around
+    // this is depending on version of GraphQL (heavily), and any change in error message format
+    // will result in wrong responses
+    private String fixMessage(String message) {
+        if (!message.startsWith("Validation error")) {
+            return message;
+        }
+        /*
+        What we get (first) and what we want (second)
+        Validation error (FieldUndefined@[allHeroes/weaknesses]) : Field 'weaknesses' in type 'SuperHero' is undefined
+        Validation error of type FieldUndefined: Field 'weaknesses' in type 'SuperHero' is undefined @ 'allHeroes/weaknesses'
+         */
+        /*
+        1: type of error (FieldUndefined)
+        2: path (allHeroes/weaknesses)
+        3: Message (Field 'weaknesses' in type 'SuperHero' is undefined)
+         */
+        Matcher matcher = VALIDATION_ERROR_PATTERN.matcher(message);
+        if (matcher.matches()) {
+            String fixedMessage = "Validation error of type " + matcher.group(1)
+                    + ": " + matcher.group(3)
+                    + " @ '" + matcher.group(2) + "'";
+
+            if (fixedMessage.contains("Literal value not in allowable values for enum")) {
+             /*
+             What we get (first) and what we want (second)
+             Validation error (WrongType@[createNewHero]) : argument 'hero.tshirtSize' with value 'EnumValue{name='XLTall'}'
+                is not a valid 'ShirtSize' - Literal value not in allowable values for enum 'ShirtSize'
+                - 'EnumValue{name='XLTall'}'
+             Validation error of type WrongType: argument 'hero.tshirtSize' with value 'EnumValue{name='XLTall'}' is not a valid
+                'ShirtSize' - Expected enum literal value not in allowable values
+                -  'EnumValue{name='XLTall'}'. @ 'createNewHero'
+             */
+                // enum failures
+                return ERROR_ENUM_REPLACE_PATTERN.matcher(fixedMessage)
+                        .replaceAll("Expected enum literal value not in allowable values -  $1.");
+            }
+
+            /*
+            actual graphql message
+            fixed graphql message
+            expected graphql message
+            Validation error (WrongType@[updateItemPowerLevel]) : argument 'powerLevel' with value
+                    'StringValue{value='Unlimited'}' is not a valid 'Int' -
+                    Expected an AST type of 'IntValue' but it was a 'StringValue'
+            Validation error of type WrongType: argument 'powerLevel' with value 'StringValue{value='Unlimited'}'
+                    is not a valid 'Int' - Expected an AST type of 'IntValue' but it
+                    was a 'StringValue' @ 'updateItemPowerLevel'
+            Validation error of type WrongType: argument 'powerLevel' with value 'StringValue{value='Unlimited'}'
+                    is not a valid 'Int' - Expected AST type 'IntValue' but
+                    was 'StringValue'. @ 'updateItemPowerLevel
+             */
+            if (fixedMessage.contains("type WrongType")) {
+                return INVALID_TYPE_REPLACE_PATTERN.matcher(fixedMessage)
+                        .replaceAll("- Expected AST type $1 but was $2. $3");
+            }
+
+            return fixedMessage;
+        } else {
+            return message;
+        }
     }
 
     @SuppressWarnings("unchecked")

--- a/microprofile/graphql/server/src/main/java/io/helidon/microprofile/graphql/server/DataFetcherUtils.java
+++ b/microprofile/graphql/server/src/main/java/io/helidon/microprofile/graphql/server/DataFetcherUtils.java
@@ -44,13 +44,12 @@ import java.util.TreeSet;
 import java.util.UUID;
 import java.util.function.Supplier;
 
-import graphql.schema.GraphQLFieldDefinition;
-
 import io.helidon.graphql.server.ExecutionContext;
 
 import graphql.GraphQLException;
 import graphql.schema.DataFetcher;
 import graphql.schema.DataFetchingEnvironment;
+import graphql.schema.GraphQLFieldDefinition;
 import graphql.schema.PropertyDataFetcher;
 import graphql.schema.PropertyDataFetcherHelper;
 import jakarta.enterprise.inject.spi.CDI;

--- a/microprofile/graphql/server/src/main/java/io/helidon/microprofile/graphql/server/DataFetcherUtils.java
+++ b/microprofile/graphql/server/src/main/java/io/helidon/microprofile/graphql/server/DataFetcherUtils.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2020, 2023 Oracle and/or its affiliates.
+ * Copyright (c) 2020, 2024 Oracle and/or its affiliates.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -42,6 +42,9 @@ import java.util.Map;
 import java.util.Set;
 import java.util.TreeSet;
 import java.util.UUID;
+import java.util.function.Supplier;
+
+import graphql.schema.GraphQLFieldDefinition;
 
 import io.helidon.graphql.server.ExecutionContext;
 
@@ -192,7 +195,7 @@ class DataFetcherUtils {
 
             // retrieve the map and return the collection of V
             Map<?, V> map = (Map<?, V>) PropertyDataFetcherHelper
-                    .getPropertyValue(propertyName, source, environment.getFieldType(), environment);
+                    .getPropertyValue(propertyName, source, environment.getFieldType(), () -> environment);
             return map.values();
         };
     }
@@ -452,7 +455,7 @@ class DataFetcherUtils {
      * An implementation of a {@link PropertyDataFetcher} which returns a formatted number.
      */
     @SuppressWarnings("rawtypes")
-    static class NumberFormattingDataFetcher extends PropertyDataFetcher {
+    static class NumberFormattingDataFetcher extends PropertyDataFetcher<Object> {
 
         /**
          * {@link NumberFormat} to format with.
@@ -483,6 +486,12 @@ class DataFetcherUtils {
         }
 
         @Override
+        public Object get(GraphQLFieldDefinition fieldDefinition, Object source,
+                          Supplier<DataFetchingEnvironment> environmentSupplier) throws Exception {
+            return formatNumber(super.get(environmentSupplier.get()), isScalar, numberFormat);
+        }
+
+        @Override
         public Object get(DataFetchingEnvironment environment) {
             return formatNumber(super.get(environment), isScalar, numberFormat);
         }
@@ -492,7 +501,7 @@ class DataFetcherUtils {
      * An implementation of a {@link PropertyDataFetcher} which returns a formatted date.
      */
     @SuppressWarnings({ "rawtypes" })
-    static class DateFormattingDataFetcher extends PropertyDataFetcher {
+    static class DateFormattingDataFetcher extends PropertyDataFetcher<Object> {
 
         /**
          * {@link DateTimeFormatter} to format with.
@@ -519,11 +528,14 @@ class DataFetcherUtils {
                 // must be java.util.Date
                 simpleDateFormat = new SimpleDateFormat(valueFormat);
             }
-            System.err.println("DateFormattingDataFetcher: propertyName=" + propertyName
-                                        + ", type=" + type
-                                        + ", valueFormat=" + valueFormat
-                                        + ", dateTimeFormatter=" + dateTimeFormatter
-                                        + ", simpleDateFormat=" + simpleDateFormat);
+        }
+
+        @Override
+        public Object get(GraphQLFieldDefinition fieldDefinition, Object source,
+                          Supplier<DataFetchingEnvironment> environmentSupplier) throws Exception {
+                return dateTimeFormatter != null
+                    ? formatDate(super.get(environmentSupplier.get()), dateTimeFormatter)
+                    : formatDate(super.get(environmentSupplier.get()), simpleDateFormat);
         }
 
         @Override

--- a/microprofile/tests/tck/tck-graphql/src/test/resources/arquillian.xml
+++ b/microprofile/tests/tck/tck-graphql/src/test/resources/arquillian.xml
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
 
-    Copyright (c) 2020, 2023 Oracle and/or its affiliates.
+    Copyright (c) 2020, 2024 Oracle and/or its affiliates.
 
     Licensed under the Apache License, Version 2.0 (the "License");
     you may not use this file except in compliance with the License.
@@ -21,7 +21,7 @@
         xmlns="http://jboss.org/schema/arquillian"
         xsi:schemaLocation="
         http://jboss.org/schema/arquillian
-        http://jboss.org/schema/arquillian/arquillian_1_0.xsd">
+        https://jboss.org/schema/arquillian/arquillian_1_0.xsd">
 
     <engine>
         <property name="deploymentExportPath">target/deployments</property>

